### ISSUE TITLE
Cyclic reference for top-level definitions of recursive match types

### ISSUE THREAD

**Issue by flomebul** [lampepfl/dotty#6362]
The following exemple no more accepted
```scala
  type LeafElem[X] = X match {
      case String => Char
      case Array[t] => LeafElem[t]
      case Iterable[t] => LeafElem[t]
      case AnyVal => X
    }
```
with the following error
```scala
-- Error: recursive.scala:11:7 -------------------------------------------------
11 |  type LeafElem[X] = X match {
   |       ^
   |illegal cyclic reference: match ... (caught cyclic reference) ... of      |type LeafElem refers back to the type itself
```

**Comment by DReigada** [comment]
This seems to still be happening for top level definitions of recursive match types

**Comment by flomebul** [comment]
For information, look like the code is accepted by dotr,but not by dotc.

**Comment by kubukoz** [comment]
Indeed, it seems like the issue is top level definitions + recursive match types. I just wasted some time fighting it :P

**Comment by flomebul** [comment]
Having a look to the regression "tests/pos/6362.scala", my feeling is that it does not cover the case. Is it possible to rewrite-it without the "object Test" encapsulation, so that the "match type" is really at top level?

**Comment by smarter** [comment]
Fixed by #9001.
